### PR TITLE
feat: UDF container should crash on exceptions

### PR DESCRIPTION
**Pull request by kohlisid**
**Requirement**
On any exception received in the user code or the execution make the UDF container restart.

**Use Case**

- This can help to fix any transient issues that might be from the code, or if any exceptions are propagated from the UDF.
- The user can raise an exception explicitly to error out, as a platform we should not consume it.

**Flow** 
For all UDF types apart from Reduce, Reducestreamer we do not expect the Numa container to crash when the UDF container crashes. 

- For Reduce/Reduce streamer
  UDF crashes due to error -> signals numa to panic -> both restart -> replay of records

- For others
  UDF crashes due to error -> signals error to numa -> UDF restarts -> replay of message due to infinite retry

**Testing** 

- Manual testing by raising explicit exceptions from UDF code
- Unit testing for error scenarios

**Testing logs**
UDF container logs
```
2024-05-01 07:44:26 CRITICAL UDFError, re-raising the error
Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/task_manager.py", line 198, in __invoke_reduce
    _ = await new_instance(keys, request_iterator, output, md)
  File "/app/example.py", line 39, in handler
    sys.exit(1)
SystemExit: 1
CRITICAL:pynumaflow._constants:Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/task_manager.py", line 198, in __invoke_reduce
    _ = await new_instance(keys, request_iterator, output, md)
  File "/app/example.py", line 39, in handler
    sys.exit(1)
SystemExit: 1
2024-05-01 07:44:26 CRITICAL 1
CRITICAL:pynumaflow._constants:1
2024-05-01 07:44:26 CRITICAL Reduce Streaming Error: AbortError('Locally aborted.')
Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/task_manager.py", line 210, in process_input_stream
    async for request in request_iterator:
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/async_servicer.py", line 27, in datum_generator
    async for d in request_iterator:
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 607, in _async_message_receiver
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 132, in read
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 82, in grpc._cython.cygrpc.RPCState.raise_for_termination
grpc._cython.cygrpc.AbortError: Locally aborted.
CRITICAL:pynumaflow._constants:Reduce Streaming Error: AbortError('Locally aborted.')
Traceback (most recent call last):
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/task_manager.py", line 210, in process_input_stream
    async for request in request_iterator:
  File "/opt/pysetup/.venv/lib/python3.10/site-packages/pynumaflow/reducestreamer/servicer/async_servicer.py", line 27, in datum_generator
    async for d in request_iterator:
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 607, in _async_message_receiver
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 132, in read
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 82, in grpc._cython.cygrpc.RPCState.raise_for_termination
grpc._cython.cygrpc.AbortError: Locally aborted.
INFO:pynumaflow._constants:Killing process: Got exception SystemExit(1)
2024-05-01 07:44:26 INFO     Killing process: Got exception SystemExit(1)
Killed
```

Numa Container logs
```
2024/04/30 07:56:22 failed ReduceFn stream.Recv(): Retryable: SystemExit(1)
2024/04/30 07:56:22 failed ReduceFn stream.Recv(): Retryable: SystemExit(1)
{"level":"panic","ts":"2024-04-30T07:56:22.400626709Z","logger":"numaflow.ReduceUDF-processor","caller":"pnf/pnf.go:155","msg":"Got an error while invoking ApplyReduce{error 26 0  gRPC client.ReduceFn failed, Retryable: SystemExit(1)}","pipeline":"even-odd-sum","vertex":"compute-sum","stacktrace":"github.com/numaproj/numaflow/pkg/reduce/pnf.(*ProcessAndForward).invokeUDF\n\t/Users/skohli/numa/numaflow/pkg/reduce/pnf/pnf.go:155"}
panic: Got an error while invoking ApplyReduce{error 26 0  gRPC client.ReduceFn failed, Retryable: SystemExit(1)}

goroutine 208 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x0?, 0x0?, {0x0?, 0x0?, 0xc0007a6c40?})
        /Users/skohli/go/pkg/mod/go.uber.org/zap@v1.26.0/zapcore/entry.go:196 +0x54
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc0007e89c0, {0x0, 0x0, 0x0})
        /Users/skohli/go/pkg/mod/go.uber.org/zap@v1.26.0/zapcore/entry.go:262 +0x24e
go.uber.org/zap.(*SugaredLogger).log(0xc0007a9318, 0x4, {0x0?, 0x0?}, {0xc0009b3ee0?, 0x0?, 0x0?}, {0x0, 0x0, 0x0})
        /Users/skohli/go/pkg/mod/go.uber.org/zap@v1.26.0/sugar.go:316 +0xec
go.uber.org/zap.(*SugaredLogger).Panic(...)
        /Users/skohli/go/pkg/mod/go.uber.org/zap@v1.26.0/sugar.go:159
github.com/numaproj/numaflow/pkg/reduce/pnf.(*ProcessAndForward).invokeUDF(0xc000520d20, {0x2a9dda0, 0xc0007aa0a0}, 0x0?, 0xc0003e13c0, {0x4047601468?, 0xc0006b6630?})
        /Users/skohli/numa/numaflow/pkg/reduce/pnf/pnf.go:155 +0x39a
created by github.com/numaproj/numaflow/pkg/reduce/pnf.(*ProcessAndForward).AsyncSchedulePnF in goroutine 1
        /Users/skohli/numa/numaflow/pkg/reduce/pnf/pnf.go:131 +0x1f9
```

**TODOs Done:**
1) Multiproc restart will work after #161 
2) Reduce sends context cancelled on restart which doesn't make the numa container restart as it retries directly on that error.
3) Async reduce error Unit test cases terminating even after mocking the method

### CODE REVIEW

**Comment by yhl25** [comment]
What is the status of this PR? is it ready for review?

File: pynumaflow/reducer/servicer/async_servicer.py [hunk 1]
```diff
@@ -101,11 +103,14 @@ async def ReduceFn(
...
             _LOGGER.critical("Reduce Error", exc_info=True)
-            context.set_code(grpc.StatusCode.UNKNOWN)
-            context.set_details(e.__str__())
-            raise e
+            # Send a context abort signal for the rpc, this is required for numa container to get
+            # the correct grpc error
+            await asyncio.gather(
+                context.abort(grpc.StatusCode.UNKNOWN, details=repr(e)), return_exceptions=True
+            )
+            terminate_on_stop(err=repr(e), parent=False)
```
Comment (vigith) [inline pynumaflow/reducer/servicer/async_servicer.py:113]: why not reuse/extend `exit_on_error`?

File: pynumaflow/reducestreamer/servicer/async_servicer.py [hunk 2]
```diff
@@ -111,22 +111,35 @@ async def ReduceFn(
         try:
             async for msg in consumer:
                 # If the message is an exception, we raise the exception
-                if isinstance(msg, Exception):
-                    await handle_error(context, msg)
-                    raise msg
+                if isinstance(msg, BaseException):
+                    handle_error(context, msg)
```
Comment (vigith) [inline pynumaflow/reducestreamer/servicer/async_servicer.py:115]: why don't i see `handle_error` in a few other places?

File: tests/reduce/test_async_reduce_err.py [hunk 3]
```diff
@@ -109,71 +108,74 @@ async def start_server(udfs):
...
-        except Exception as err:
-            self.assertTrue(
-                "reduce create operation error: invalid number of windows" in err.__str__()
-            )
-            return
-        self.fail("Expected an exception.")
-
-    def __stub(self):
-        return reduce_pb2_grpc.ReduceStub(_channel)
+# TODO: Check why terminating even after mocking
```
Comment (vigith) [inline tests/reduce/test_async_reduce_err.py:111]: please fix this.

File: pynumaflow/reducer/servicer/async_servicer.py (hunk 1 above)
Comment (kohlisid) [reply pynumaflow/reducer/servicer/async_servicer.py:113]: `exit_on_error` internally calls `terminate_on_stop`
The difference being that `exit_on_error` manipulates the context before exiting. 
Here we were already aborting the context, hence I was calling terminate directly.

If required, I can extend the functionality to one function itself using a flag

File: pynumaflow/reducestreamer/servicer/async_servicer.py (hunk 2 above)
Comment (kohlisid) [reply pynumaflow/reducestreamer/servicer/async_servicer.py:115]: `handle_error` is used in reducestreamer only as in this function we extract the traceback from the exception before logging it out. 
In other places we log out the exception along with the traceback option directly.

But in reducestreamer we write the exceptions to the global result queue, hence while processing them we extract the traceback and then log it for better contextual logging.

File: tests/reduce/test_async_reduce_err.py (hunk 3 above)
Comment (kohlisid) [reply tests/reduce/test_async_reduce_err.py:111]: We see a behavior where the target function does not bind to the mocked value. In our other test cases I used the same approach and it was working fine. I had tried different things, but wasn't helping so had commented this for now. 
I discussed this with @ab93 as well today. We would need to understand this further to see where is the problem stemming from.

Comment (vigith) [reply tests/reduce/test_async_reduce_err.py:111]: I think we have to figure that out before merging the PR, right?

Comment (kohlisid) [reply tests/reduce/test_async_reduce_err.py:111]: Ideally shouldn't be a blocker, manually have confirmed that the restarts occur as expected for both containers. This is only for UT.

Comment (vigith) [reply tests/reduce/test_async_reduce_err.py:111]: can you open a new issue to fix it and prioritize that?

Comment (kohlisid) [reply tests/reduce/test_async_reduce_err.py:115]: @vigith I have it working! Updated the mocks in some functions! 
The CI is passing now

I have still left the TODO there for myself to check and understand this a bit more deeper.

**Review by vigith — APPROVED** [review]

File: pynumaflow/mapstreamer/servicer/async_servicer.py [hunk 4]
```diff
@@ -33,26 +34,35 @@ async def MapStreamFn(
...
+                    event_time=request.event_time.ToDatetime(),
+                    watermark=request.watermark.ToDatetime(),
+                    headers=dict(request.headers),
+                ),
+                context,
+            ):
+                yield mapstream_pb2.MapStreamResponse(result=res)
+        except BaseException as err:
+            _LOGGER.critical("UDFError, re-raising the error", exc_info=True)
+            exit_on_error(context, str(err))
```
Comment (ab93) [inline pynumaflow/mapstreamer/servicer/async_servicer.py:52]: Let's use repr for errors whenever possible

**Review by ab93 — APPROVED** [review]

**Pull request merged**